### ISSUE TITLE
NLP-BERT - Enhance speed for run_pretraining.py

### ISSUE THREAD

**Issue by abccabrera** [tensorflow/models#9307]
# Prerequisites

Please answer the following question for yourself before submitting an issue.

- [ /] I checked to make sure that this feature has not been requested already.

## 1. The entire URL of the file you are using

https://github.com/tensorflow/models/blob/master/official/nlp/bert/run_pretraining.py

## 2. Describe the feature you request

Enhance performance of pretraining of BERT.

## 3. Additional context

Running the script takes approximately 7 hours for 1,000 steps.  
The original BERT is pretrained with  500,000 to 1 million steps; which at this current rate training for new models may take too much time to complete.

script I ran:
!python models/official/nlp/bert/run_pretraining.py \
    --input_files=gs://data5709-bert/eurusd1.tfrecord \
    --model_dir=gs://data5709-bert/testmodel5TF2/ \
    --bert_config_file=config.json \
    --train_batch_size=16 \
    --max_seq_length=128 \
    --max_predictions_per_seq=20 \
    --num_train_epochs=10 \
    --num_steps_per_epoch=20000 \
    --warmup_steps=10000 \
    --learning_rate=2e-5 \
    --tpu=$TPU_NAME

## 4. Are you willing to contribute it? (Yes or No)
I'm quite new with tensorflow, but yes if I can, I will.

**Comment by saberkun** [comment]
Are you using gpus and how many gpus are you using? For gpus, you need to enable mix precision to speed up.
We recommend at least using 8 V100. There are some padding for the inputs at this moment and we will improve that. For the best GPU setting, @zongweiz may have more inputs.
It is also good to try TPUs and the TPU-v3-32 is a decent choice to finish the training faster.

**Comment by zongweiz** [comment]
As @saberkun pointed out, mixed precision training could speed up your workload. Try adding --dtype = 'fp16', --loss_scale = 'dynamic', --num_gpus=8, --train_batch_size=(total batch size on 8GPU).  Here is a quick guideline about mixed precision training: https://www.tensorflow.org/guide/mixed_precision

You could also try out the new experimental XLA compilation feature. XLA could help you use higher batch size and also achieve higher throughput.
You need to add tf.function(experimental_compile=True) to your training step and optimizer weight update (as we have shown in MLPerf 0.7 submission):
training step: https://github.com/mlperf/training_results_v0.7/blob/master/Google/benchmarks/bert/implementations/bert-cloud-TF2.0-gpu-v100-8/tf2_common/modeling/model_training_utils.py#L299-L316 
optimizer weight update: https://github.com/mlperf/training_results_v0.7/blob/master/Google/benchmarks/bert/implementations/bert-cloud-TF2.0-gpu-v100-8/optimization.py#L358-L364

**Comment by saberkun** [comment]
@abccabrera According to your command line, it looks like a few gpus are used. It is far from enough to really train the bert model.
